packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -683,8 +875,22 @@ class TextPainter {
   ///
   /// Each box corresponds to a [PlaceholderSpan] in the order they were defined
   /// in the [InlineSpan] tree.
-  List<TextBox>? get inlinePlaceholderBoxes => _inlinePlaceholderBoxes;
-  List<TextBox>? _inlinePlaceholderBoxes;
+  List<TextBox>? get inlinePlaceholderBoxes {
+    final _TextPainterLayoutCacheWithOffset? layout = _layoutCache;
+    if (layout == null) {
+      return null;
+    }
+    final List<TextBox> rawBoxes = layout._inlinePlaceholderBoxes
+                               ??= layout.paragraph.getBoxesForPlaceholders();
```

Comment:
Maybe it would be a bit cleaner if the caching were to happen inside the `_TextPainterLayoutCacheWithOffset`? The `layout._inlinePlaceholderBoxes` getter would then either consult with the paragraph to get the placeholder or return the cached values.

Reply:
(same for other places that access "cached" values)